dds/idl/annotations.h

```diff
@@ -76,105 +76,136 @@ class Annotation {
   AST_Annotation_Decl* declaration_;
 };
 
-AST_Expression::AST_ExprValue* get_annotation_member_ev(
-  AST_Annotation_Appl* appl, const char* member_name);
+AST_Expression::AST_ExprValue* get_annotation_member_ev(AST_Annotation_Appl* appl,
+                                                        const char* member_name);
 
-bool get_bool_annotation_member_value(
-  AST_Annotation_Appl* appl, const char* member_name);
+bool get_bool_annotation_member_value(AST_Annotation_Appl* appl,
+                                      const char* member_name,
+                                      const bool* default_value);
 
-ACE_UINT32 get_u32_annotation_member_value(
-  AST_Annotation_Appl* appl, const char* member_name);
+ACE_UINT32 get_u32_annotation_member_value(AST_Annotation_Appl* appl,
+                                           const char* member_name,
+                                           const ACE_UINT32* default_value);
+
+std::string get_str_annotation_member_value(AST_Annotation_Appl* appl,
+                                            const char* member_name,
+                                            const std::string* default_value);
+
+/**
+ * Annotation that logically provide a value when absent.
+ */
+template <typename T>
+class AbsentValue {
+public:
+  AbsentValue(const T& value)
+    : absent_value(value)
+  {}
+
+  const T absent_value;
+};
 
 /**
  * Annotation with a Single Member Named "value"
  */
 template <typename T>
 class AnnotationWithValue : public Annotation {
 public:
-  /**
-   * Default value if the node DOESN'T have the annotation. This is different
-   * than the default value in the annotation definition.
-   */
-  virtual T default_value() const
-  {
-    return T();
-  }
-
   /**
    * If node has the annotation, this sets value to the annotation value and
    * returns true.
-   * If node does not have the annotation, this sets value to default_value()
+   * If node does not have the annotation, this sets value to absent_value()
    * and returns false.
    */
   virtual bool node_value_exists(AST_Decl* node, T& value) const
   {
     AST_Annotation_Appl* appl = find_on(node);
-    value = value_from_appl(appl);
-    return appl;
-  }
+    if (!appl) { return false; }
```

Comment:
This isn't setting a value like it should (or at least what the comment says).

Reply:
That was a commenting fail.

Reply:
This still isn't consistent with the comment.

Reply:
Shouldn't it be setting the absent value?

Reply:
If a field has a `@key` annotation, will it have a parameter list where the first parameter is a boolean true?  If so, all of the default stuff can go away.

How would you reword the comment to make it clearer?

Reply:
> If a field has a @key annotation, will it have a parameter list where the first parameter is a boolean true? If so, all of the default stuff can go away.

Yes. An `AST_Annotation_Appl` will always have values for its fields, coming from either the defaults in the annotation definition or the annotation application. If an annotation application in IDL doesn't define all the non-defaulted fields of the definition, then it's a compiler error.

> How would you reword the comment to make it clearer?

The comment is saying it's setting the default value if the annotation isn't there, but it just returns if the node doesn't have the annotation. It used to be done by `value_from_appl`, but that's not happening anymore. I'm still not sure why you did that, but either the part in the comment has to be fixed or preferably you have the function set `value` when  the annotation is missing.